testing/trino-testing/src/main/java/io/trino/testing/AbstractLocalDynamicFilteringTest.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testing;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Ints;
+import io.trino.Session;
+import io.trino.execution.QueryStats;
+import io.trino.operator.OperatorStats;
+import io.trino.spi.QueryId;
+import io.trino.sql.planner.OptimizerConfig;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.planner.OptimizerConfig.JoinDistributionType.BROADCAST;
+import static io.trino.testing.DataProviders.toDataProvider;
+import static org.testng.Assert.assertEquals;
+
+public abstract class AbstractLocalDynamicFilteringTest
+        extends AbstractTestQueryFramework
+{
+    private static final int ORDERS_COUNT = 15000;
+    private static final int CUSTOMER_COUNT = 1500;
+
+    protected Map<String, String> getPropertyMap()
+    {
+        // Adjust DF limits to test edge cases
+        return ImmutableMap.of(
+                "dynamic-filtering.small-broadcast.max-distinct-values-per-driver", "100",
+                "dynamic-filtering.small-broadcast.range-row-limit-per-driver", "100",
+                "dynamic-filtering.large-broadcast.max-distinct-values-per-driver", "100",
+                "dynamic-filtering.large-broadcast.range-row-limit-per-driver", "100000");
+    }
+
+    @Test
+    public void testCrossJoinLargeBuildSideDynamicFiltering()
+    {
+        // Probe-side is fully scanned because the build-side is too large for dynamic filtering:
+        assertDynamicFiltering(
+                "SELECT * FROM orders o, customer c WHERE o.custkey < c.custkey AND c.name < 'Customer#000001000' AND o.custkey > 1000",
+                noJoinReordering(BROADCAST),
+                0,
+                ORDERS_COUNT, CUSTOMER_COUNT);
+    }
+
+    @Test(timeOut = 30_000, dataProvider = "joinDistributionTypes")
+    public void testJoinDynamicFilteringMultiJoin(OptimizerConfig.JoinDistributionType joinDistributionType)
```

Comment:
It's actually not about local dynamic filtering, but about how scheduler schedules stages (and it doesn't lock). So it's not local DF test. This maybe isn't the best place for it, but I couldn't find a better one